### PR TITLE
Update PETSc to 3.21.4 and fixup breaking tests

### LINKED ISSUE #27227: Update PETSc submodule
### Motivation

PETSc has recently updated to version 3.21.1. MOOSE should be kept up-to-date with this release.

EDIT (May 10 2023): Changed to 3.21.1

### Design

Update the PETSc submodule and corresponding packages to 3.21.1.

### Impact

New features and bugfixes from PETSc.

### PR DESCRIPTION
**Pull request by cticenhour**
Closes #27227

Replaces #27604 and #27260.

Still have some optimization modules tests to fix, but might as well kick off testing to make sure I fixed up the others across all platforms.

Apps to fix:
- [x] [Griffin](https://github.inl.gov/ncrc/griffin/pull/2022)
- [x] [Subchannel](https://github.inl.gov/ncrc/subchannel/pull/331) This is like a half-check because I (@lindsayad) only got 1 of the 3 exodiffs from CIVET locally. I am also running with PETSc 3.21.3 tho
- [x] [BlackBear](https://github.com/idaholab/blackbear/pull/399)
- [x] [Ferret](https://github.com/mangerij/ferret/pull/339)
- [x] [SAM](https://git-nse.egs.anl.gov/SAM/SAM/-/merge_requests/945)
- [x] [BISON](https://github.inl.gov/ncrc/bison/pull/6067)
- [x] [Pronghorn](https://github.inl.gov/ncrc/pronghorn/pull/1275)
- [x] [Grizzly](https://github.inl.gov/ncrc/grizzly/pull/774)

### CODE REVIEW

**Comment by nmnobre** [comment]
I don't think I have permissions to push to your branch, but I think if we `git cherry-pick` farscape-project/moose@c64b2f6, the PETSc install checks should pass.

**Comment by nmnobre** [comment]
What a pain in the arse this turned out to be, is it always like this? 😅

The Apptainer check that we are now skipping failed only on the MUMPS test and we're now seeing the same "double free or corruption" error in some tests using LU (which I think is provided by MUMPS...). We could patch PETSc to downgrade MUMPS but that's not very pretty, is it? 🫠

**Comment by cticenhour** [comment]
> What a pain in the arse this turned out to be, is it always like this? 😅

Becoming more and more common with these updates, unfortunately. This MUMPS issue is a new one, but I've had more issues out of TAO (leading to issues in the optimization module) on every recent update.

> We could patch PETSc to downgrade MUMPS but that's not very pretty, is it? 🫠

I should check to see if we have latitude in what MUMPS version we can select via configure.

**Comment by lindsayad** [comment]
`--download-mumps-commit=commitid` should do it

**Comment by cticenhour** [comment]
Thanks Alex 😄 I had found this on Friday, but had not had the chance to get the commit hash and add this yet. Would rather like not having to pin MUMPS, but needs must I guess....

**Comment by nmnobre** [comment]
> Thanks Alex 😄 I had found this on Friday, but had not had the chance to get the commit hash and add this yet. Would rather like not having to pin MUMPS, but needs must I guess....

`help.addArgument('MUMPS', '-download-mumps-avoid-mpi-in-place', nargs.ArgBool(None, 0, 'Let MUMPS not use MPI_IN_PLACE. Since MUMPS-5.6.2, it can be used to avoid a bug in MPICH older than 4.0b1'))`

as per https://gitlab.com/petsc/petsc/-/blob/0062f871983a63f49e6f0d2ab37969d3e8074e10/config/BuildSystem/config/packages/MUMPS.py#L26.

Can we also move to PETSc 3.21.2?

**Comment by cticenhour** [comment]
@nmnobre Alex tried this and it did not work for us, unfortunately. See https://gitlab.com/petsc/petsc/-/issues/1569#note_1921209585 for context.

I will be updating to 3.21.2 once I address one more failing modules test. Should be able to (finally) get this in this week.

**Comment by nmnobre** [comment]
> @nmnobre Alex tried this and it did not work for us, unfortunately. See https://gitlab.com/petsc/petsc/-/issues/1569#note_1921209585 for context.

Ah... dang it :/

File: scripts/configure_petsc.sh [hunk 1]
```diff
@@ -78,6 +78,15 @@ function configure_petsc()
     fi
   fi
 
+  # Patch for PETSc-TAO in order to make TaoSolve more thread-safe. This should be removed when the
+  # patch contents have been added to a PETSc release.
+  # See: https://gitlab.com/petsc/petsc/-/merge_requests/7047
```
Comment (cticenhour) [inline scripts/configure_petsc.sh:83]:
```suggestion
  # See context of patch content here: https://gitlab.com/petsc/petsc/-/merge_requests/7047
```

**Comment by lindsayad** [comment]
alright with #26848 in, this is the next exciting stack PR

**Comment by lindsayad** [comment]
this is an important PR we need for some NEAMS M2 tasks. Let me know if there's anything I can do to help

**Comment by cticenhour** [comment]
The most time consuming item will be the app patches - if you can assist with that, I'd be grateful. I think there are a couple more modules failures that popped up as I was fixing the others I knew of too. It'll be a hard week to focus on this one while I am on travel.

**Comment by lindsayad** [comment]
sounds good, I will look at the apps

**Comment by lindsayad** [comment]
I've added a list in the OP to track app progress

File: conda/libmesh/meta.yaml [hunk 2]
```diff
@@ -4,7 +4,7 @@
 #
 # As well as any directions pertaining to modifying those files.
 # ALSO: Follow the directions in scripts/tests/versioner_hashes.yaml
-{% set build = 0 %}
+{% set build = 2 %}
```
Comment (milljm) [inline conda/libmesh/meta.yaml:7]:
```suggestion
{% set build = 1 %}
```

File: conda/moose-dev/conda_build_config.yaml [hunk 3]
```diff
@@ -3,8 +3,8 @@ mpi:
   - openmpi
 
 moose_libmesh:
-  - moose-libmesh 2024.07.27 mpich_0
-  - moose-libmesh 2024.07.27 openmpi_0
+  - moose-libmesh 2024.07.27 mpich_2
+  - moose-libmesh 2024.07.27 openmpi_2
```
Comment (milljm) [inline conda/moose-dev/conda_build_config.yaml:7]:
```suggestion
  - moose-libmesh 2024.07.27 mpich_1
  - moose-libmesh 2024.07.27 openmpi_1
```

File: conda/moose/conda_build_config.yaml [hunk 4]
```diff
@@ -3,7 +3,7 @@ mpi:
   - openmpi
 
 moose_dev:
-  - moose-dev 2024.07.27
+  - moose-dev 2024.07.30
```
Comment (milljm) [inline conda/moose/conda_build_config.yaml:6]:
```suggestion
  - moose-dev 2024.08.05
```

Comment (milljm) [reply conda/moose/conda_build_config.yaml:6]: (This is what's breaking "Conda MOOSE (Linux)"

File: conda/petsc/meta.yaml [hunk 5]
```diff
@@ -8,7 +8,7 @@
 # As well as any directions pertaining to modifying those files.
 # ALSO: Follow the directions in scripts/tests/versioner_hashes.yaml
 {% set build = 1 %}
```
Comment (milljm) [inline conda/petsc/meta.yaml:10]:
```suggestion
{% set build = 0 %}
```

File: framework/doc/packages_config.yml [hunk 6]
```diff
@@ -10,8 +10,8 @@ maximum_python: 3.11
 mpich: 4.2.1
 openmpi: 4.1.6
 petsc_alt: 3.11.4
-petsc: 3.20.3
-moose_dev: 2024.07.27
+petsc: 3.21.4
+moose_dev: 2024.07.30
```
Comment (milljm) [inline framework/doc/packages_config.yml:14]:
```suggestion
moose_dev: 2024.08.05
```

**Review by milljm — CHANGES_REQUESTED** [review]
Just some version bumping fixes

**Comment by gsgall** [comment]
> @gsgall this is triggering the Zapdos test to fail again 😦

I think I have a potential fix for this with Zapdos PR [#255](https://github.com/shannon-lab/zapdos/pull/255)

**Comment by joshuahansel** [comment]
THM changes approved 👍

File: apptainer/petsc.def [hunk 7]
```diff
@@ -90,6 +90,8 @@ From: {{ APPTAINER_FROM }}
 %files
     {{ MOOSE_DIR }}/scripts/configure_petsc.sh {{ ROOT_BUILD_DIR }}/scripts/configure_petsc.sh
     {{ MOOSE_DIR }}/scripts/update_and_rebuild_petsc.sh {{ ROOT_BUILD_DIR }}/scripts/update_and_rebuild_petsc.sh
+    {{ MOOSE_DIR }}/scripts/apple-silicon-hdf5-autogen.patch {{ ROOT_BUILD_DIR }}/scripts/apple-silicon-hdf5-autogen.patch
```
Comment (loganharbour) [inline apptainer/petsc.def:93]: I don't think this one is needed for linux, is it?

File: apptainer/petsc.def [hunk 8]
```diff
@@ -90,6 +90,8 @@ From: {{ APPTAINER_FROM }}
...
+    {{ MOOSE_DIR }}/scripts/tao_restore_viewer.patch {{ ROOT_BUILD_DIR }}/scripts/tao_restore_viewer.patch
```
Comment (loganharbour) [inline apptainer/petsc.def:94]: I think both of these should be added to the influential files

File: conda/mpi/build.sh [hunk 9]
```diff
@@ -16,7 +16,7 @@ mkdir -p "${PREFIX}/etc/conda/activate.d" "${PREFIX}/etc/conda/deactivate.d"
 cat <<EOF > "${PREFIX}/etc/conda/activate.d/activate_${PKG_NAME}.sh"
 export CC=mpicc CXX=mpicxx FC=mpif90 F90=mpif90 F77=mpif77 C_INCLUDE_PATH=${PREFIX}/include \
  MOOSE_NO_CODESIGN=true MPIHOME=${PREFIX} CXXFLAGS="$ACTIVATION_CXXFLAGS" HDF5_DIR=${PREFIX} \
- FI_PROVIDER=tcp
+ FI_PROVIDER=tcp OMPI_MCA_mca_base_component_show_load_errors=0
```
Comment (loganharbour) [inline conda/mpi/build.sh:19]: This only needs to be set for openmpi, right?

File: conda/petsc/build.sh [hunk 10]
```diff
@@ -6,6 +6,7 @@ export PETSC_ARCH=arch-conda-c-opt
 # Remove std=C++17 from CXXFLAGS as we specify the C++ dialect for PETSc as C++17 in configure_petsc.
 # Specifying both causes an error as of PETSc 3.17.
 CXXFLAGS=${CXXFLAGS//-std=c++[0-9][0-9]}
+LDFLAGS=${LDFLAGS//-Wl,-dead_strip_dylibs}
```
Comment (loganharbour) [inline conda/petsc/build.sh:9]: Do we have a listed reason for this one?

File: apptainer/petsc.def (hunk 7 above)
Comment (lindsayad) [reply apptainer/petsc.def:93]: no

File: conda/mpi/build.sh (hunk 9 above)
Comment (lindsayad) [reply conda/mpi/build.sh:19]: yes. And the FI_PROVIDER option only needs to be set for mpich

File: conda/petsc/build.sh (hunk 10 above)
Comment (lindsayad) [reply conda/petsc/build.sh:9]: can add

File: conda/mpi/build.sh (hunk 9 above)
Comment (milljm) [reply conda/mpi/build.sh:19]: is it harmless either way? Or should we only set it for one variant or another?

File: apptainer/petsc.def (hunk 8 above)
Comment (milljm) [reply apptainer/petsc.def:94]: @lindsayad do you know how to do this? It's easy peasy. Just add the paths to the `moose/scripts/versioner.yaml` file under the appropriate 'influential' file list for petsc.

File: apptainer/petsc.def (hunk 7 above)
Comment (lindsayad) [reply apptainer/petsc.def:93]: how do you optionally add files based on architecture?

Comment (milljm) [reply apptainer/petsc.def:93]:
```bash
if [[ "${mpi}" == 'mpich' ]]; then
  do mpich related stuff
else
  do openmpi related stuff
fi
```

The 'keys' being set in `conda_build_config.yaml` become available as variables. So this section:
```yaml
mpi:
  - mpich
  - openmpi
```
controls what `mpi`'s value is at the time.

File: conda/mpi/build.sh (hunk 9 above)
Comment (lindsayad) [reply conda/mpi/build.sh:19]: I'm not quite sure what would be a good way to detect mpich vs. openmpi within this script

Comment (milljm) [reply conda/mpi/build.sh:19]: The way to do this, is how I mistakenly instructed how to do it in another comment on this PR above...

**Comment by lindsayad** [comment]
oh wow, all green! 🚂

**Review by milljm — APPROVED** [review]
Looks good to me!

**Comment by lindsayad** [comment]
@loganharbour let me know if your conversations are resolved

File: conda/mpi/build.sh (hunk 9 above)
Comment (loganharbour) [reply conda/mpi/build.sh]: I'm fine with ignoring this. It's harmless

**Review by loganharbour — APPROVED** [review]

**Pull request merged**